File: python/Binary_Search/day10/leastCapacityShip.py
def shippingOfDays(weight, cap):
    n = len(weight)
    days = 1
    load = 0

    for i in range(n):
        if load + weight[i] > cap:
            days = days + 1
            load = weight[i]
        
        else:
            load += weight[i]
        
    return days



def sumOfTotal(weight):
    total = 0

    for num in weight:
        total += num
    
    return total



def leastCapacityDays(weight, days):
    n = len(weight)
    maxi = float('-inf')

    for i in range(n):
        maxi = max(maxi, weight[i])
    

    low = maxi
    high = sumOfTotal(weight)

    while low <= high:
        mid = (low + high) // 2

        noOfDays = shippingOfDays(weight, mid)

        if noOfDays <= days:
            high = mid - 1
        
        else:
            low = mid + 1
    
    return low



if __name__ == "__main__":
    n = int(input("Enter number of elements: "))

    print("Enter the weights:", end=" ")
    weight = list(map(int, input().split()))

    days = int(input("Enter days: "))

    ans = leastCapacityDays(weight, days)
    print(f"Minimum ship capcity needed = {ans}")

File: java/Binary_Search/Day10/LeastCapacityShip.java
package Day10;
import java.util.*;

public class LeastCapacityShip {
    public int shippingOfDays(int[] weight, int cap) {
        int days = 1;
        int load = 0;

        for (int i = 0; i < weight.length; i++) {
            if (load + weight[i] > cap) {
                days++;
                load = weight[i];
            }
            else {
                load += weight[i];
            }
        }

        return days;
    }



    // helper function to calculate total sum...
    public int sumOfTotal(int[] weight) {
        int total = 0;
        for (int num : weight) total += num;

        return total;
    }


    public int leastCapacityDays(int[] weight, int days) {
        int n = weight.length;
        int maxi = Integer.MIN_VALUE;

        for (int i = 0; i < n; i++) {
            maxi = Math.max(maxi, weight[i]);
        }

        int low = maxi;
        int high = sumOfTotal(weight);

        while (low <= high) {
            int mid = (low + high) / 2;
            
            int noOfDays = shippingOfDays(weight, mid);

            if (noOfDays <= days) {
                high = mid - 1;
            }
            else {
                low = mid + 1;
            }
        }

        return low;
    }


    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();

        int[] weight = new int[n];
        System.out.print("Enter the weights: ");
        for (int i = 0; i < n; i++) {
            weight[i] = sc.nextInt();
        }

        System.out.print("Enter days: ");
        int days = sc.nextInt();

        LeastCapacityShip obj = new LeastCapacityShip();
        int ans = obj.leastCapacityDays(weight, days);
        System.out.println("Minimum ship capacity needed = " + ans + " " + "days.");


        sc.close();
    }
}